src/Storage/Notmuch.hs

```diff
@@ -36,20 +36,19 @@ messageToMail
     => T.Text
     -> Message
     -> IO NotmuchMail
-messageToMail ignoredTag m =
+messageToMail ignoredTag m = do
+    tgs <- tags m
     NotmuchMail <$>
-    (decodeUtf8 . fromMaybe "" <$> messageHeader "Subject" m) <*>
-    (decodeUtf8 . fromMaybe "" <$> messageHeader "To" m) <*>
-    (decodeUtf8 . fromMaybe "" <$> messageHeader "From" m) <*>
-    messageFilename m <*>
-    messageDate m <*>
-    (tagsToText m ignoredTag) <*>
-    isNewMail m ignoredTag
+      (decodeUtf8 . fromMaybe "" <$> messageHeader "Subject" m) <*>
+      (decodeUtf8 . fromMaybe "" <$> messageHeader "To" m) <*>
+      (decodeUtf8 . fromMaybe "" <$> messageHeader "From" m) <*>
+      messageFilename m <*>
+      messageDate m <*>
+      (tagsToText tgs ignoredTag) <*>
+      isNewMail tgs ignoredTag
 
-tagsToText :: HasTags a => a -> T.Text -> IO [T.Text]
-tagsToText m ignored = do
-  t <- tags m
-  pure $ filter (/= ignored) $ decodeUtf8 <$> t
+tagsToText :: [Tag] -> T.Text -> IO [T.Text]
```

Comment:
This should be a pure function.  If the caller want to lift it into `IO` let
them apply `pure` at the call site.